// @dojo/stores/patch/createJsonPointer.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    function navigate(path, target) {
        return path.segments().reduce(function (prev, next) {
            return prev ? prev[next] : prev;
        }, target);
    }
    exports.navigate = navigate;
    function decode(segment) {
        return segment.replace(/~1/g, '/').replace(/~0/g, '~');
    }
    function encode(segment) {
        return segment.replace(/~/g, '~0').replace(/\//g, '~1');
    }
    function toString() {
        var segments = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            segments[_i] = arguments[_i];
        }
        return segments.reduce(function (prev, next) {
            return prev + '/' + encode(next);
        });
    }
    function createJsonPointer() {
        var segments = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            segments[_i] = arguments[_i];
        }
        return {
            segments: function () {
                return segments.map(function (segment) { return decode(segment); });
            }, toString: function () {
                return toString.apply(void 0, segments);
            },
            push: function (segment) {
                return createJsonPointer.apply(void 0, segments.concat(segment));
            },
            pop: function () {
                return createJsonPointer.apply(void 0, segments.slice(0, segments.length - 1));
            }
        };
    }
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.default = createJsonPointer;
});
