// @dojo/widgets/main.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./createProjector", "./components/button/createButton", "./components/textinput/createTextInput", "./createWidgetBase", "./d"], factory);
    }
})(function (require, exports) {
    "use strict";
    var createProjector_1 = require("./createProjector");
    exports.createProjector = createProjector_1.default;
    var createButton_1 = require("./components/button/createButton");
    exports.createButton = createButton_1.default;
    var createTextInput_1 = require("./components/textinput/createTextInput");
    exports.createTextInput = createTextInput_1.default;
    var createWidgetBase_1 = require("./createWidgetBase");
    exports.createWidgetBase = createWidgetBase_1.default;
    var d_1 = require("./d");
    exports.v = d_1.v;
    exports.w = d_1.w;
});
