// @dojo/stores/query/createStoreRange.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    function serializeRange(range) {
        return "limit(" + range.count + (range.start ? ',' + range.start : '') + ")";
    }
    function createRange(start, count, serializer) {
        return {
            apply: function (data) {
                return data.slice(start, start + count);
            },
            queryType: 2 /* Range */,
            toString: function (rangeSerializer) {
                return (rangeSerializer || serializer || serializeRange)(this);
            },
            start: start,
            count: count,
            incremental: false
        };
    }
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.default = createRange;
});
